File: crates/agent/src/components/artifacts.rs
//! THE WORDS THE SHELF IS WRITTEN IN: what this group has produced, who each
//! piece is for, and whether anything confirmed it is there.
//!
//! It is not a `Component`. Artifacts are a FACULTY (`crate::faculty::artifact`),
//! which declares the block — id, slot, intent, stability — and
//! `components::Sensed` renders whatever a host most recently wrote for it.
//! What lives here is the VOCABULARY, in one place, exactly as `space_parts`
//! holds the space's (`crates/agent/src/components/space.rs:63`) and
//! `memory_parts` holds memory's.
//!
//! **NO DURABILITY SENTENCE HERE, AND NO `WorkspacePort` ON THE SENSE.** That
//! wording belongs to `components::space`, which renders whenever there IS a
//! folder to describe, and `crates/agent/src/environment/mod.rs:166-173` records
//! that a TEST decided that rather than an argument. One truth, one wording, one
//! place; a second copy of it here would be the drift that note exists to stop.

use context::{text, Part};

use crate::artifact::{Artifact, Shelf, SHELF_LIMIT, READ_ARTIFACT, RECORD_ARTIFACT};
use crate::toolbox::Toolbox;

/// The parts a host leaves in `AgentState.senses["artifacts"]` for
/// `components::Sensed` to render.
///
/// EMPTY when the shelf is empty, and that is the whole rule: emptiness becomes
/// `Fidelity::Elided` (`assemble` starts a partless section there,
/// `crates/context/src/assemble.rs:110`), so a group that has produced nothing
/// gets no heading and no blank section rather than a paragraph saying so.
/// Every capability may be absent (I15), and this is how the paper spells it.
///
/// `tools` is the agent's RESOLVED toolbox, and it is here for the reason
/// `space_parts` takes one: the closing line names calls by name, and a name
/// the agent was never granted advertises a capability that is not there — the
/// one failure I15 forbids. An agent holding neither call is told what is on
/// the shelf and offered nothing, which is the true description of its
/// situation.
pub fn artifact_parts(shelf: &Shelf, tools: &Toolbox) -> Vec<Part> {
    if shelf.items.is_empty() {
        return Vec::new();
    }
    let mut lines: Vec<String> = shelf.items.iter().take(SHELF_LIMIT).map(line).collect();
    // I16: a shelf that quietly showed twenty of twenty-five would be the paper
    // lying about what the group has. The count is the cheapest true sentence.
    if let Some(hidden) = shelf.items.len().checked_sub(SHELF_LIMIT).filter(|n| *n > 0) {
        lines.push(format!("…and {hidden} more on this shelf, not listed here."));
    }
    if let Some(closing) = offered(tools) {
        lines.push(closing);
    }
    text(lines.join("\n"))
}

/// One artifact, one line. The `uri` is not rendered: it is `artifact://<space>/
/// <name>` by construction and the space is already named one block up, so
/// printing it spends budget restating two things the model was just told.
///
/// `(unconfirmed)` is the gate this whole faculty's header argues for — the
/// port's real answer, or an honest note that nothing in the recording thread
/// could look. A size claimed for an artifact nobody measured is exactly the
/// sentence I16 exists to forbid.
fn line(artifact: &Artifact) -> String {
    let size = match artifact.bytes {
        Some(n) => format!("{n} bytes"),
        None => "unconfirmed".to_string(),
    };
    format!(
        "- {} ({}, rev {}, {size}) for {} — {} [by {}]",
        artifact.name,
        artifact.kind,
        artifact.revision,
        artifact.audience,
        artifact.description,
        artifact.by
    )
}

/// The closing line: only the calls this agent actually holds, and nothing at
/// all when it holds neither. A pair rather than a name alone because the clause
/// cannot be spelled out of the name — `components::space`'s `READERS` rule.
fn offered(tools: &Toolbox) -> Option<String> {
    let held: Vec<&str> = CALLS
        .iter()
        .filter(|(name, _)| tools.get(name).is_some())
        .map(|(_, clause)| *clause)
        .collect();
    match held.is_empty() {
        true => None,
        false => Some(held.join(" ")),
    }
}

const CALLS: [(&str, &str); 2] = [
    (READ_ARTIFACT, "read_artifact opens any of these by name."),
    (RECORD_ARTIFACT, "record_artifact puts one more on this shelf."),
];

File: packages/agent/src/faculty/artifacts.js
/**
 * THE ARTIFACT SHELF — what this group has PRODUCED, as opposed to what it has
 * said. It is not part of `space` for three reasons, and the third is why it
 * gets its own block rather than three more lines in the space's: a note falls
 * off the board at a limit and a deliverable may not; the shelf is declarable
 * on its own, so a read-only agent with a folder and no shelf is representable
 * (I6, default deny); and `## space` answers where the group works while this
 * answers what came out of the work and who it is for.
 *
 * Two calls and no more. `write_file` already writes the file and `list_files`
 * already lists the folder — what the group had no way to say was WHICH of
 * those files is a deliverable and what it is for. A lister or a deleter here
 * would be a second answer to a question the workspace tools have answered.
 * @module
 */

import { arg, tool } from '../tools.js'

/** The faculty's name, its block's id, and the key a host writes its rendered parts under. */
export const ARTIFACTS = 'artifacts'

/**
 * Slot 57, between the space it belongs to (55) and the clock (60). The gaps of
 * ten in `SLOT` exist for exactly this, so nothing is renumbered.
 */
const ARTIFACTS_SLOT = 57

/** @type {import('./index.js').Faculty} */
export const artifactsFaculty = {
  name: ARTIFACTS,
  block: {
    id: ARTIFACTS,
    slot: ARTIFACTS_SLOT,
    intent: 'What this group has produced that outlives a turn, and who each piece is for.',
    stability: 'semi_static',
  },
  // `record_artifact` does not promise the file was CHECKED — nothing here can
  // check it — so it claims only what it does: the record reaches everyone.
  tools: [
    tool({
      name: 'record_artifact',
      description: 'Put a file you have written on this space\'s shelf, so every agent working '
        + 'here sees it named and described in their prompt without reading it. Recording the '
        + 'same name again replaces the entry.',
      args: [
        arg('name', 'string', 'the file\'s path in the workspace folder'),
        arg('description', 'string', 'one line saying what it is'),
        arg('audience', 'string', 'who it is for', { required: false }),
      ],
      mutates: true,
      needs: 'space',
    }),
    tool({
      name: 'read_artifact',
      description: 'Read an artifact on this space\'s shelf by its name. For a big one, add '
        + '\'offset\' and \'limit\' — whole numbers of BYTES — to read one window of it; the '
        + 'answer states the whole file\'s size, so you can ask for the rest.',
      args: [
        arg('name', 'string', 'the artifact\'s name, as it appears on the shelf'),
        arg('offset', 'number', 'the byte to start at', { required: false }),
        arg('limit', 'number', 'how many bytes to read', { required: false }),
      ],
      needs: 'space',
    }),
  ],
}
